File: src/npmpc/dynamics/integrators/midpoint.py
from npmpc.dynamics.integrators import IntegrationBase
import torch
import casadi as ca


class MidpointIntegration(IntegrationBase):

    def step(self, x: torch.Tensor, u: torch.Tensor, dt: float, p: torch.Tensor=None):
        k1 = self.dynamics.dynamics(x, u, p)
        k2 = self.dynamics.dynamics(x + dt/2*k1, u, p)
        return x + dt*k2


    def integrate(self, x0: torch.Tensor, u: torch.Tensor, dt: float, p: torch.Tensor=None):
        x = x0.unsqueeze(-2).repeat((*([1]*len(x0.shape[:-1])), u.shape[-2]+1, 1))
        for i in range(u.shape[-2]):
            x[...,i+1,:] = self.step(x[...,i,:], u[...,i,:], dt, p)
        return x


    def casadi_explicit(self, x0: ca.SX, u: ca.SX, dt: float, p: ca.SX=None):
        x = ca.SX.zeros(u.size1()+1, x0.size2())
        x[0] = x0[0]
        for i in range(u.size1()):
            k1 = self.dynamics.casadi_dynamics(x[i,:], u[i,:], p)
            k2 = self.dynamics.casadi_dynamics(x[i,:] + dt/2*k1, u[i,:], p)
            x[i+1,:] = x[i,:] + dt*k2
        return x


    def casadi_implicit(self, x: ca.MX, u: ca.MX, dt: float, p: ca.MX=None):
        constraints = []
        for i in range(u.size1()):
            constraints.append(x[i+1,:] == x[i,:] + dt*self.dynamics.casadi_dynamics((x[i,:]+x[i+1,:])/2, u[i,:], p))
        return constraints

File: src_cpp/npmpc/dynamics/integrators/midpoint.cpp
#include "npmpc/dynamics/integrators/midpoint.hpp"

#include "npmpc/dynamics/furuta.hpp"

namespace npmpc::dynamics::integrators {

std::vector<casadi::MX> MidpointIntegration::casadiImplicit(
        const casadi::MX& x, const casadi::MX& u, double dt, const casadi::DM& p) const {
    std::vector<casadi::MX> constraints;
    constraints.reserve(static_cast<size_t>(u.size1()));

    for (casadi_int i = 0; i < u.size1(); ++i) {
        casadi::MX xi = x(i, casadi::Slice());
        casadi::MX xNext = x(i + 1, casadi::Slice());
        casadi::MX xMid = (xi + xNext) / 2;
        constraints.push_back(xNext == xi + dt * FurutaDynamics::casadiDynamics(xMid, u(i, casadi::Slice()), p));
    }
    return constraints;
}

} // namespace npmpc::dynamics::integrators
